src/common/quantile.h

```diff
@@ -444,150 +233,124 @@ struct WXQSummary : public WQSummary<DType, RType> {
-   * \param src source summary
-   * \param maxsize size we can afford in the pruned sketch
+   * \brief set current summary to be merged summary of sa and sb
+   * \param sa first input summary to be merged
+   * \param sb second input summary to be merged
    */
-  inline void SetPrune(const GKSummary &src, size_t maxsize) {
-    if (src.size <= maxsize) {
-      this->CopyFrom(src); return;
-    }
-    const RType max_rank = src.MaxRank();
-    this->size = maxsize;
-    data[0] = src.data[0];
-    size_t n = maxsize - 1;
-    RType top = 1;
-    for (size_t i = 1; i < n; ++i) {
-      RType k = (i * max_rank) / n;
-      while (k > src.data[top + 1].rmax) ++top;
-      // assert src.data[top].rmin <= k
-      // because k > src.data[top].rmax >= src.data[top].rmin
-      if ((k - src.data[top].rmin) < (src.data[top+1].rmax - k)) {
-        data[i] = src.data[top];
-      } else {
-        data[i] = src.data[top + 1];
-      }
-    }
-    data[n] = src.data[src.size - 1];
-  }
-  inline void SetCombine(const GKSummary &sa,
-                         const GKSummary &sb) {
+  void SetCombine(const WXQSummary &sa, const WXQSummary &sb) {
     if (sa.size == 0) {
-      this->CopyFrom(sb); return;
+      this->CopyFrom(sb);
+      return;
     }
     if (sb.size == 0) {
-      this->CopyFrom(sa); return;
+      this->CopyFrom(sa);
+      return;
     }
-    CHECK(sa.size > 0 && sb.size > 0) << "invalid input for merge";
+    CHECK(sa.size > 0 && sb.size > 0);
     const Entry *a = sa.data, *a_end = sa.data + sa.size;
     const Entry *b = sb.data, *b_end = sb.data + sb.size;
-    this->size = sa.size + sb.size;
+    // extended rmin value
     RType aprev_rmin = 0, bprev_rmin = 0;
     Entry *dst = this->data;
     while (a != a_end && b != b_end) {
-      if (a->value < b->value) {
-        *dst = Entry(bprev_rmin + a->rmin,
-                     a->rmax + b->rmax - 1, a->value);
-        aprev_rmin = a->rmin;
-        ++dst; ++a;
+      // duplicated value entry
+      if (a->value == b->value) {
+        *dst = Entry(a->rmin + b->rmin, a->rmax + b->rmax, a->wmin + b->wmin,
+                     a->value);
+        aprev_rmin = a->RMinNext();
+        bprev_rmin = b->RMinNext();
+        ++dst;
+        ++a;
+        ++b;
+      } else if (a->value < b->value) {
+        *dst = Entry(a->rmin + bprev_rmin, a->rmax + b->RMaxPrev(), a->wmin,
+                     a->value);
+        aprev_rmin = a->RMinNext();
+        ++dst;
+        ++a;
       } else {
-        *dst = Entry(aprev_rmin + b->rmin,
-                     b->rmax + a->rmax - 1, b->value);
-        bprev_rmin = b->rmin;
-        ++dst; ++b;
+        *dst = Entry(b->rmin + aprev_rmin, b->rmax + a->RMaxPrev(), b->wmin,
+                     b->value);
+        bprev_rmin = b->RMinNext();
+        ++dst;
+        ++b;
       }
     }
     if (a != a_end) {
-      RType bprev_rmax = (b_end - 1)->rmax;
+      RType brmax = (b_end - 1)->rmax;
```

Comment:
Please change it to a more verbose name.

Reply:
I actually have no idea what any of this code does, it's just copy pasted. I will spend a bit more time learning the algorithms to see if I can clean it up more.